ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleStorageBackend.scala

```diff
@@ -0,0 +1,520 @@
+      |  values ({deduplicationKey}, {deduplicateUntil})""".stripMargin
+
+  def upsertDeduplicationEntry(
+      key: String,
+      submittedAt: Instant,
+      deduplicateUntil: Instant,
+  )(connection: Connection): Int =
+    SQL(SQL_INSERT_COMMAND)
+      .on(
+        "deduplicationKey" -> key,
+        "submittedAt" -> submittedAt,
+        "deduplicateUntil" -> deduplicateUntil,
+      )
+      .executeUpdate()(connection)
+
+  override def batch(dbDtos: Vector[DbDto]): AppendOnlySchema.Batch =
+    OracleSchema.schema.prepareData(dbDtos)
+
+  override def insertBatch(connection: Connection, batch: AppendOnlySchema.Batch): Unit =
+    OracleSchema.schema.executeUpdate(batch, connection)
+
+  def commandCompletions(
+      startExclusive: Offset,
+      endInclusive: Offset,
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  )(connection: Connection): List[CompletionStreamResponse] =
+    TemplatedStorageBackend.commandCompletions(
+      startExclusive = startExclusive,
+      endInclusive = endInclusive,
+      applicationId = applicationId,
+      submittersInPartiesClause = arrayIntersectionWhereClause("submitters", parties),
+    )(connection)
+
+  def activeContractWithArgument(readers: Set[Ref.Party], contractId: ContractId)(
+      connection: Connection
+  ): Option[StorageBackend.RawContract] =
+    TemplatedStorageBackend.activeContractWithArgument(
+      treeEventWitnessesWhereClause = arrayIntersectionWhereClause("tree_event_witnesses", readers),
+      contractId = contractId,
+    )(connection)
+
+  def activeContractWithoutArgument(readers: Set[Ref.Party], contractId: ContractId)(
+      connection: Connection
+  ): Option[String] =
+    TemplatedStorageBackend.activeContractWithoutArgument(
+      treeEventWitnessesWhereClause = arrayIntersectionWhereClause("tree_event_witnesses", readers),
+      contractId = contractId,
+    )(connection)
+
+  def contractKey(readers: Set[Ref.Party], key: Key)(
+      connection: Connection
+  ): Option[ContractId] =
+    TemplatedStorageBackend.contractKey(
+      flatEventWitnesses = columnPrefix =>
+        arrayIntersectionWhereClause(s"$columnPrefix.flat_event_witnesses", readers),
+      key = key,
+    )(connection)
+
+  def transactionsEventsSingleWildcardParty(
+      startExclusive: Long,
+      endInclusive: Long,
+      party: Ref.Party,
+      limit: Option[Int],
+      fetchSizeHint: Option[Int],
+  )(connection: Connection): Vector[EventsTable.Entry[Raw.FlatEvent]] =
+    TemplatedStorageBackend.transactionsEventsSingleWildcardParty(
+      startExclusive = startExclusive,
+      endInclusive = endInclusive,
+      party = party,
+      partyArrayContext = partyArrayContext,
+      witnessesWhereClause = arrayIntersectionWhereClause("flat_event_witnesses", Set(party)),
+      limitExpr = limitClause(limit),
+      fetchSizeHint = fetchSizeHint,
+      submittersInPartiesClause = arrayIntersectionWhereClause("submitters", Set(party)),
```

Comment:
Previously, this used array equality, not array intersection. Would it be difficult to write a function that checks whether an array column contains exactly one given element?

In practice, the `transactionsEventsSingleWildcardParty` / `transactionsEventsSinglePartyWithTemplates` queries should be used much more often than `transactionsEventsOnlyWildcardParties` / `transactionsEventsSameTemplates`, as the latter are for the rarely used multi-party subscription feature.

Reply:
outdated again